File: 심수연/3주차/260115.py
# https://www.acmicpc.net/problem/1654

import sys
input = sys.stdin.readline

K, N = map(int, input().split())

lan_cable = []

for _ in range(K):
  lan_cable.append(int(input()))
  
start = 1
end = max(lan_cable)

answer = 0

while start <= end:
  mid = (start + end) // 2 # 자를 랜선의 길이
  
  count = 0 # 랜선 개수
  
  for i in lan_cable:
    count += i // mid
    if count >= N:
      break
  
  if count >= N: # count 가 N보다 크거나 같으면
    answer = mid # 자를 랜선의 길이를 answer 에 담기
    start = mid + 1 # 자를 랜선의 길이를 더 키우기
  else:
    end = mid - 1 # 자를 랜선의 길이를 더 줄이기

print(answer)
    
    
  

File: 서정우/3주차/260115.js
// https://www.acmicpc.net/problem/33524

const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "../input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);
const problems = input[1].split(" ").map(Number);
const users = input[2].split(" ").map(Number);

const sizes = [0];
for (let i = 1; ; i++) {
  const size = 3 * i * (i - 1) + 1;
  if (size > N) break;
  sizes.push(size);
}

problems.sort((a, b) => a - b);

const upperBound = (arr, target) => {
  let l = 0;
  let r = arr.length;
  while (l < r) {
    const mid = (l + r) >> 1;
    if (arr[mid] <= target) l = mid + 1;
    else r = mid;
  }
  return l;
};

const result = users.map((userSkill) => {
  const count = upperBound(problems, userSkill);
  if (count === 0) return 0;

  let left = 0;
  let right = sizes.length - 1;
  let max = 0;

  while (left <= right) {
    const mid = (left + right) >> 1;
    if (sizes[mid] <= count) {
      max = mid;
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return max;
});

console.log(result.join(" "));
